encoding/string_encode.go

```diff
@@ -1,30 +1,182 @@
+)
+
+// Default initial size of data slice.
+const (
+	defaultInitalDataSliceSize = 1 << 8
+)
+
+// Maximum cardinality allowed for dictionary encoding.
+var (
+	dictEncodingMaxCardinality = 1 << 16
 )
 
-// StringEncoder encodes string values.
+// StringEncoder encodes string valuesIt.
 type StringEncoder interface {
 	// Encode encodes a collection of strings and writes the encoded bytes to the writer.
 	// Callers should explicitly call `Reset` before subsequent call to `Encode`.
-	Encode(writer io.Writer, values ForwardStringIterator) error
-
-	// Reset resets the encoder.
-	Reset()
+	Encode(writer io.Writer, valuesIt RewindableStringIterator) error
 }
 
 // StringEnc is a string encoder.
-type StringEnc struct{}
+type StringEnc struct {
+	buf             []byte
+	metaProto       encodingpb.StringMeta
+	dictionaryProto encodingpb.StringArray
+	data            []string
+}
 
 // NewStringEncoder creates a new string encoder.
-func NewStringEncoder(writer io.Writer) *StringEnc { return &StringEnc{} }
+func NewStringEncoder() *StringEnc {
+	return &StringEnc{
+		buf: make([]byte, 8),
+		metaProto: encodingpb.StringMeta{
+			// Default compression type.
+			Compression: encodingpb.CompressionType_ZSTD,
+		},
+		data: make([]string, defaultInitalDataSliceSize),
+	}
+}
 
 // Encode encodes a collection of strings and writes the encoded bytes to the writer.
-func (enc *StringEnc) Encode(writer io.Writer, values ForwardStringIterator) error {
-	return errors.New("not implemented")
+func (enc *StringEnc) Encode(
+	writer io.Writer,
+	valuesIt RewindableStringIterator,
+) error {
+	// TODO(bodu): Do some perf benchmarking to see whether we want to allocate a new map
+	// or clear an existing one.
+	dictionary := make(map[string]int64)
+	var (
+		idx       int64
+		curr      string
+		maxLength int
+	)
+	for valuesIt.Next() {
+		curr = valuesIt.Current()
+		if len(dictionary) < dictEncodingMaxCardinality {
+			// Only add to dictionary if
+			if _, ok := dictionary[curr]; !ok {
+				dictionary[curr] = idx
+				// Ensure data size or grow slice. Target size is idx+1.
+				enc.data = enc.ensureStringSliceSize(enc.data, int(idx)+1)
+				enc.data[idx] = curr
+				idx++
+			}
+		}
+		maxLength = int(math.Max(float64(len(curr)), float64(maxLength)))
+	}
+
+	if err := valuesIt.Err(); err != nil {
+		return err
+	}
+
+	// Rewind values.
+	valuesIt.Rewind()
+
+	// TODO(bodu): This should take into account the total # of items
+	// at some point. The total count should exceed the # of uniques by a certain
+	// margin to justify table compression.
+	if len(dictionary) >= dictEncodingMaxCardinality {
+		enc.metaProto.Encoding = encodingpb.EncodingType_RAW_SIZE
+	} else {
+		enc.metaProto.Encoding = encodingpb.EncodingType_DICTIONARY
+	}
+
+	enc.buf = bytes.EnsureBufferSize(enc.buf, enc.metaProto.Size(), bytes.DontCopyData)
+	if err := ProtoEncode(&enc.metaProto, enc.buf, writer); err != nil {
+		return err
+	}
+
+	var compressWriter io.Writer
```

Comment:
Probably better to call this `writer` and rename `w` below to `compressWriter`.

Reply:
👍